rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -352,17 +362,46 @@ class IntraProcessManager
       auto subscription_base = subscription_it->second.lock();
       if (subscription_base) {
         auto subscription = std::dynamic_pointer_cast<
-          rclcpp::experimental::SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>
+          rclcpp::experimental::SubscriptionIntraProcessBuffer<PublishedType,
+          PublishedTypeAllocator, PublishedTypeDeleter, ROSMessageType>
           >(subscription_base);
         if (nullptr == subscription) {
-          throw std::runtime_error(
-                  "failed to dynamic cast SubscriptionIntraProcessBase to "
-                  "SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>, which "
-                  "can happen when the publisher and subscription use different "
-                  "allocator types, which is not supported");
+          auto ros_message_subscription = std::dynamic_pointer_cast<
+            rclcpp::experimental::ROSMessageIntraProcessBuffer<ROSMessageType,
+            ROSMessageTypeAllocator, ROSMessageTypeDeleter>
+            >(subscription_base);
+
+          if (nullptr == ros_message_subscription) {
+            throw std::runtime_error(
+                    "failed to dynamic cast SubscriptionIntraProcessBase to "
+                    "SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>, which "
```

Comment:
this is not the correct failure.
We definitely failed to do that,but we also failed the `ros_message_subscription` cast.

The same applies to the runtime error in the following function